src/testrender/cuda/optix_raytracer.cu

```diff
@@ -34,6 +45,251 @@ __device__ __constant__ RenderParams render_params;
+    sg.Ng = normalize(optixTransformNormalFromObjectToWorldSpace(V3_TO_F3(local_sg.Ng)));
+    sg.P  = V3_TO_F3(P.val());
+    sg.dPdx        = V3_TO_F3(P.dx());
+    sg.dPdy        = V3_TO_F3(P.dy());
+    sg.dPdu        = local_sg.dPdu;
+    sg.dPdv        = local_sg.dPdv;
+    sg.u           = local_sg.u;
+    sg.v           = local_sg.v;
+    sg.Ci          = nullptr;
+    sg.surfacearea = local_sg.surfacearea;
+    sg.backfacing  = dot(V3_TO_F3(sg.N), V3_TO_F3(sg.I)) > 0.0f;
+    sg.shaderID    = local_sg.shaderID;
+
+    if (sg.backfacing) {
+        sg.N  = -sg.N;
+        sg.Ng = -sg.Ng;
+    }
+
+    sg.raytype        = raytype;
+    sg.flipHandedness = dot(V3_TO_F3(sg.N), cross(V3_TO_F3(sg.dPdx), V3_TO_F3(sg.dPdy))) < 0.0f;
+}
+
+
+static inline __device__ void
+execute_shader(ShaderGlobalsType& sg, char* closure_pool)
+{
+    if (sg.shaderID < 0) {
+        // TODO: should probably never get here ...
+        return;
+    }
+
+    // Pack the "closure pool" into one of the ShaderGlobals pointers
+    *(int*)&closure_pool[0] = 0;
+    sg.renderstate          = &closure_pool[0];
+
+    // Create some run-time options structs. The OSL shader fills in the structs
+    // as it executes, based on the options specified in the shader source.
+    NoiseOptCUDA noiseopt;
+    TextureOptCUDA textureopt;
+    TraceOptCUDA traceopt;
+
+    // Pack the pointers to the options structs in a faux "context",
+    // which is a rough stand-in for the host ShadingContext.
+    ShadingContextCUDA shading_context = { &noiseopt, &textureopt, &traceopt };
+    sg.context                         = &shading_context;
+
+    // Run the OSL callable
+    void* interactive_ptr = reinterpret_cast<void**>(
+        render_params.interactive_params)[sg.shaderID];
+    const unsigned int shaderIdx = 2u + sg.shaderID + 0u;
+    optixDirectCall<void, ShaderGlobalsType*, void*, void*, void*, int, void*>(
+        shaderIdx, &sg /*shaderglobals_ptr*/,
+        nullptr /*groupdata_ptr*/,
+        nullptr /*userdata_base_ptr*/,
+        nullptr /*output_base_ptr*/,
+        0 /*shadeindex - unused*/,
+        interactive_ptr /*interactive_params_ptr*/
+    );
+}
+
+
+static __device__ Color3
+eval_background(const Dual2<Vec3>& dir, void* /*ctx*/, int bounce = -1)
+{
+    ShaderGlobalsType sg;
+    memset((char*)&sg, 0, sizeof(ShaderGlobalsType));
+    sg.I    = V3_TO_F3(dir.val());
+    sg.dIdx = V3_TO_F3(dir.dx());
+    sg.dIdy = V3_TO_F3(dir.dy());
+    if (bounce >= 0)
+        sg.raytype = bounce > 0 ? Ray::DIFFUSE : Ray::CAMERA;
+    sg.shaderID = render_params.bg_id;
+
+    alignas(8) char closure_pool[256];
+    execute_shader(sg, closure_pool);
+    return process_background_closure((const ClosureColor*)sg.Ci);
+}
+
+
+// Return a direction towards a point on the sphere
+// Adapted from Sphere::sample in ../raytracer.h
+static __device__ float3
+sample_sphere(const Vec3& x, const SphereParams& sphere, float xi, float yi,
+              float& pdf)
```

Comment:
Any chance we can call the original code to avoid duplication? Even if creating a temporary Sphere, the compiler should optimize out cruft.

Reply:
That is a possibility I had considered. I'll see how well it works in practice.